x-pack/apm-server/aggregation/servicemetrics/aggregator.go

```diff
@@ -204,45 +209,101 @@ func (a *Aggregator) processTransaction(event *model.APMEvent) model.APMEvent {
+		m:       make(map[uint64][]*metricsMapEntry),
+		space:   make([]metricsMapEntry, maxSize),
 	}
 }
 
+type metricsMapEntry struct {
+	serviceMetrics
+	aggregationKey
+}
+
 func (mb *metricsBuffer) storeOrUpdate(key aggregationKey, metrics serviceMetrics, logger *logp.Logger) bool {
+	// hash does not use the serviceMetrics so it is safe to call concurrently.
+	hash := key.hash()
+
+	// Full lock because serviceMetrics cannot be updated atomically.
 	mb.mu.Lock()
 	defer mb.mu.Unlock()
-	old, ok := mb.m[key]
-	if !ok {
-		n := len(mb.m)
-		half := mb.maxSize / 2
-		if !ok {
-			switch n {
-			case mb.maxSize:
-				return false
-			case half - 1:
-				logger.Warn("service metrics groups reached 50% capacity")
-			case mb.maxSize - 1:
-				logger.Warn("service metrics groups reached 100% capacity")
+
+	entries, ok := mb.m[hash]
+	if ok {
+		for offset, old := range entries {
+			if old.aggregationKey.equal(key) {
+				entries[offset].serviceMetrics = serviceMetrics{
+					transactionDuration: old.transactionDuration + metrics.transactionDuration,
+					transactionCount:    old.transactionCount + metrics.transactionCount,
+					failureCount:        old.failureCount + metrics.failureCount,
+					successCount:        old.successCount + metrics.successCount,
+				}
+				return true
 			}
 		}
+	} else if mb.entries >= len(mb.space) {
+		return false
 	}
-	mb.m[key] = serviceMetrics{
-		transactionDuration: old.transactionDuration + metrics.transactionDuration,
-		transactionCount:    old.transactionCount + metrics.transactionCount,
-		failureCount:        old.failureCount + metrics.failureCount,
-		successCount:        old.successCount + metrics.successCount,
+
+	half := mb.maxSize / 2
+	if !ok {
+		switch mb.entries {
+		case mb.maxSize:
+			return false
+		case half - 1:
+			logger.Warn("service metrics groups reached 50% capacity")
+		case mb.maxSize - 1:
+			logger.Warn("service metrics groups reached 100% capacity")
+		}
 	}
+
+	entry := &mb.space[mb.entries]
+	entry.aggregationKey = key
+
+	entry.serviceMetrics = serviceMetrics{
+		transactionDuration: metrics.transactionDuration,
+		transactionCount:    metrics.transactionCount,
+		failureCount:        metrics.failureCount,
+		successCount:        metrics.successCount,
+	}
+
+	mb.m[hash] = append(entries, entry)
+	mb.entries++
 	return true
 }
 
 type aggregationKey struct {
+	labels.AggregatedGlobalLabels
+	comparable
+}
+
+func (k *aggregationKey) hash() uint64 {
+	var h xxhash.Digest
+	var buf [8]byte
+	binary.LittleEndian.PutUint64(buf[:], uint64(k.timestamp.UnixNano()))
+	h.Write(buf[:])
+
+	k.AggregatedGlobalLabels.Write(&h)
+	h.WriteString(k.agentName)
+	h.WriteString(k.serviceEnvironment)
+	h.WriteString(k.serviceName)
+	h.WriteString(k.transactionType)
+	return h.Sum64()
+}
+
+func (k *aggregationKey) equal(key aggregationKey) bool {
+	return k.comparable == key.comparable && k.AggregatedGlobalLabels.Equals(&key.AggregatedGlobalLabels)
```

Comment:
nit; I think this reads slightly better on smaller editors.
```suggestion
	return k.comparable == key.comparable && 
		k.AggregatedGlobalLabels.Equals(&key.AggregatedGlobalLabels)
```